pennylane/operation.py

```diff
@@ -480,67 +480,81 @@ def matrix(self, wire_order=None):
         guaranteed.
 
-        This private method allows eigenvalues to be computed
-        directly without instantiating the operators first.
+        .. note::
+            This method gets overwritten by subclasses to define the eigenvalues
+            of a particular operator.
 
-        The default implementation relies on the presence of the
-        :attr:`compute_matrix` method.
+        Args:
+            params (list): trainable parameters of this operator, as stored in ``op.parameters``
+            hyperparams (dict): non-trainable hyperparameters of this operator, as stored in ``op.hyperparameters``
 
-        To return the eigenvalues of *instantiated* operators,
-        please use the :attr:`~.Operator.eigvals` property instead.
+        Returns:
+            array: eigenvalues
 
         **Example:**
 
-        >>> qml.RZ._eigvals(0.5)
+        >>> qml.RZ.compute_eigvals(0.5)
         array([0.96891242-0.24740396j, 0.96891242+0.24740396j])
         >>> qml.PauliX(wires=0).diagonalizing_gates()
         [Hadamard(wires=[0])]
-        >>> qml.PauliX._eigvals()
+        >>> qml.PauliX.compute_eigvals()
         array([1, -1])
-
-        Returns:
-            array: eigenvalue representation
         """
-        return np.linalg.eigvals(cls.compute_matrix(*params))
+        return NotImplementedError
 
-    @property
     def eigvals(self):
-        r"""Eigenvalues of an instantiated operator.
+        r"""Eigenvalues of the operator.
 
         If :attr:`diagonalizing_gates` are specified, the order of the
         eigenvalues needs to match the order of
         the computational basis vectors when the observable is
         diagonalized using these ops. Otherwise, no particular order is
         guaranteed.
 
+        .. note::
+            By default, this method calls the static method ``compute_eigvals``,
+            which is used by subclasses to define the actual eigenvalues. If no
+            eigenvalues are defined, it is attempted to compute them from the matrix
+            representation.
+
+        Returns:
+            array: eigenvalues
+
         **Example:**
 
         >>> U = qml.RZ(0.5, wires=1)
-        >>> U.eigvals
+        >>> U.eigvals()
         array([0.96891242-0.24740396j, 0.96891242+0.24740396j])
         >>> qml.PauliX(wires=0).diagonalizing_gates()
         [Hadamard(wires=[0])]
         >>> qml.PauliX.eigvals()
         array([1, -1])
-
-        Returns:
-            array: eigvals representation
         """
-        return self._eigvals(*self.parameters)
+
+        try:
+            return self.compute_eigvals(*self.parameters, **self.hyperparameters)
+        except NotImplementedError:
+            # By default, compute the eigenvalues from the matrix representation.
+            # This will raise a NotImplementedError if the matrix is undefined.
+            return np.linalg.eigvals(self.compute_matrix(*self.parameters, **self.hyperparameters))
```

Comment:
I suppose, on second thought, this should be?
```suggestion
            return np.linalg.eigvals(self.matrix())
```

Reply:
True, this is more elegant!